### PR TITLE
Added produces text/plain to endpoint in order to make it work again …

### PR DESCRIPTION
**Pull request by predic8**
…with prometheus scrapes.

Fixes prometheus/client_java/265

### CODE REVIEW

**Comment by brian-brazil** [comment]
Could you add a unittest so this doesn't break again?

**Comment by massenz** [comment]
Is there an ETA for this into a release? This currently breaks all integration with Prometheus, as it gets a 406 and does not report any metrics.

**Comment by brian-brazil** [comment]
Once this PR is in, I plan to do a release.

**Comment by predic8** [comment]
An Unit test for the modification ( requests with Accept header containing just text/plain) is added now. To make it work Spring and Spring Boot dependencies had to be updated and some tests had to be rewritten.

File: simpleclient_spring_boot/src/test/java/io/prometheus/client/spring/boot/PrometheusEndpointTest.java [hunk 1]
```diff
@@ -6,33 +6,33 @@
...
-import org.springframework.boot.test.SpringApplicationConfiguration;
-import org.springframework.boot.test.TestRestTemplate;
-import org.springframework.boot.test.WebIntegrationTest;
-import org.springframework.http.HttpStatus;
-import org.springframework.http.ResponseEntity;
-import org.springframework.test.context.junit4.SpringJUnit4ClassRunner;
-import org.springframework.web.client.RestTemplate;
+import org.springframework.boot.test.context.SpringBootTest;
+import org.springframework.boot.test.web.client.TestRestTemplate;
+import org.springframework.http.*;
```
Comment (brian-brazil) [inline simpleclient_spring_boot/src/test/java/io/prometheus/client/spring/boot/PrometheusEndpointTest.java]: Please keep imports explicit

File: simpleclient_common/src/main/java/io/prometheus/client/exporter/common/TextFormat.java [hunk 2]
```diff
@@ -11,6 +11,7 @@
    * Content-type for text version 0.0.4.
    */
   public final static String CONTENT_TYPE_004 = "text/plain; version=0.0.4; charset=utf-8";
+  public final static String CONTENT_TYPE = "text/plain";
```
Comment (brian-brazil) [inline simpleclient_common/src/main/java/io/prometheus/client/exporter/common/TextFormat.java]: This doesn't belong here, this is not the content type of the Prometheus text format

File: simpleclient_spring_boot/src/main/java/io/prometheus/client/spring/boot/PrometheusMvcEndpoint.java [hunk 3]
```diff
@@ -24,7 +24,7 @@ public PrometheusMvcEndpoint(PrometheusEndpoint delegate) {
 
   @RequestMapping(
           method = {RequestMethod.GET},
-          produces = {TextFormat.CONTENT_TYPE_004}
+          produces = {TextFormat.CONTENT_TYPE_004, TextFormat.CONTENT_TYPE}
```
Comment (brian-brazil) [inline simpleclient_spring_boot/src/main/java/io/prometheus/client/spring/boot/PrometheusMvcEndpoint.java]: I think what's needed here is a wildcard, as this only solves it for one content type if I understand this correctly

File: simpleclient_spring_boot/src/test/java/io/prometheus/client/spring/boot/PrometheusEndpointTest.java [hunk 4]
```diff
@@ -6,33 +6,33 @@
...
-import static org.junit.Assert.assertTrue;
+import static org.springframework.boot.test.context.SpringBootTest.WebEnvironment.RANDOM_PORT;
 
-@RunWith(SpringJUnit4ClassRunner.class)
-@SpringApplicationConfiguration(DummyBootApplication.class)
-@WebIntegrationTest(randomPort = true)
+@RunWith(SpringRunner.class)
 @EnablePrometheusEndpoint
+@SpringBootTest(webEnvironment = RANDOM_PORT, classes = DummyBootApplication.class)
+@TestPropertySource( properties = "management.security.enabled=false")
```
Comment (brian-brazil) [inline simpleclient_spring_boot/src/test/java/io/prometheus/client/spring/boot/PrometheusEndpointTest.java]: extra space after the (, and below

File: simpleclient_spring_boot/src/test/java/io/prometheus/client/spring/boot/PrometheusEndpointTest.java [hunk 5]
```diff
@@ -47,11 +47,15 @@ public void testMetricsExportedThroughPrometheusEndpoint() {
...
+
+    HttpHeaders headers = new HttpHeaders();
+    headers.set("Accept", "text/plain");
+
+    ResponseEntity<String> metricsResponse = template.exchange(getBaseUrl() + "/prometheus?name[]=foo_bar", HttpMethod.GET, new HttpEntity(headers), String.class);
 
     // then:
     assertEquals(HttpStatus.OK, metricsResponse.getStatusCode());
-    assertTrue(StringUtils.deleteWhitespace(TextFormat.CONTENT_TYPE_004).equals(metricsResponse.getHeaders().getContentType().toString()));
+    assertEquals(StringUtils.deleteWhitespace(TextFormat.CONTENT_TYPE_004),metricsResponse.getHeaders().getContentType().toString().toLowerCase());
```
Comment (brian-brazil) [inline simpleclient_spring_boot/src/test/java/io/prometheus/client/spring/boot/PrometheusEndpointTest.java]: space after comma

File: simpleclient_spring_boot/src/test/java/io/prometheus/client/spring/boot/DummyBootApplication.java [hunk 6]
```diff
@@ -9,4 +9,5 @@
 @SpringBootApplication
 class DummyBootApplication {
 
+
```
Comment (brian-brazil) [inline simpleclient_spring_boot/src/test/java/io/prometheus/client/spring/boot/DummyBootApplication.java]: why the extra blank line?

File: simpleclient_spring_boot/src/test/java/io/prometheus/client/spring/boot/PrometheusEndpointTest.java [hunk 7]
```diff
@@ -6,33 +6,33 @@
...
-import static org.junit.Assert.assertTrue;
+import static org.springframework.boot.test.context.SpringBootTest.WebEnvironment.RANDOM_PORT;
 
-@RunWith(SpringJUnit4ClassRunner.class)
-@SpringApplicationConfiguration(DummyBootApplication.class)
-@WebIntegrationTest(randomPort = true)
+@RunWith( SpringRunner.class)
 @EnablePrometheusEndpoint
+@SpringBootTest( webEnvironment = RANDOM_PORT, classes = DummyBootApplication.class)
+@TestPropertySource( properties = "management.security.enabled=false")
```
Comment (brian-brazil) [inline simpleclient_spring_boot/src/test/java/io/prometheus/client/spring/boot/PrometheusEndpointTest.java]: I meant to remove the extra space, make it consistent one way or the other.

File: simpleclient_spring_boot/src/test/java/io/prometheus/client/spring/boot/PrometheusEndpointTest.java (hunk 1 above)
Comment (brian-brazil) [inline simpleclient_spring_boot/src/test/java/io/prometheus/client/spring/boot/PrometheusEndpointTest.java]: This is still here

**Comment by massenz** [comment]
Any progress here? the way I understand things here, this client is broken until this fix is merged, isn't it?

**Comment by brian-brazil** [comment]
The client works, except for this exposition method. The other exposition methods still work, as do older versions.

**Comment by dgradl** [comment]
Anything I can do to help with this so it can be merged?   I did a bit of testing with the updated PrometheusMvcEndpoint from this PR and my prometheus server here and was able to successfully scrape my boot service.

**Comment by brian-brazil** [comment]
If you can take this PR and address the final minor comments that'd save me from doing it.

**Comment by brian-brazil** [comment]
And if you could verify #266 is resolved too that'd be great.

**Comment by predic8** [comment]
@brian-brazil the last commit addressed the minor comments already.

**Comment by brian-brazil** [comment]
My last two comments weren't addressed, did you forget to push the change?

File: simpleclient_spring_boot/src/test/java/io/prometheus/client/spring/boot/PrometheusMvcEndpointTest.java [hunk 8]
```diff
@@ -4,45 +4,55 @@
...
-import org.springframework.http.HttpEntity;
-import org.springframework.http.HttpHeaders;
-import org.springframework.http.HttpMethod;
-import org.springframework.http.HttpStatus;
-import org.springframework.http.ResponseEntity;
-import org.springframework.test.context.junit4.SpringJUnit4ClassRunner;
-import org.springframework.web.client.RestTemplate;
+import org.springframework.boot.test.context.SpringBootTest;
+import org.springframework.boot.test.web.client.TestRestTemplate;
+import org.springframework.http.*;
```
Comment (brian-brazil) [inline simpleclient_spring_boot/src/test/java/io/prometheus/client/spring/boot/PrometheusMvcEndpointTest.java]: This one too, and then we can get it merged.

**Comment by brian-brazil** [comment]
Thanks!

**Comment by dgradl** [comment]
Great!  Are you planning to release 0.0.24 soon then?

**Comment by brian-brazil** [comment]
The release is already in progress. It takes a while though.

**Comment by massenz** [comment]
I've tried to use the `0.0.24` release, and I no longer get the `406` error, but now Spring MVC raises this exception:
```
java.lang.IllegalStateException: Ambiguous handler methods mapped for HTTP path 'http://apiserver:8080/prometheus': {public org.springframework.http.ResponseEntity io.prometheus.client.spring.boot.PrometheusMvcEndpoint.value(java.util.Set), public java.lang.Object org.springframework.boot.actuate.endpoint.mvc.EndpointMvcAdapter.invoke()}
```
Not sure if this is a direct consequence of this PR, but the bottom line is, my Spring Boot app does not communicate with the Prometheus server.

**Comment by dgradl** [comment]
0.0.24 is working fine in my implementation.    

Your prometheus.yml looks similar to mine.

Are you using Spring Boot 1.5.4.RELEASE?  0.0.24 also included this upgrade according to the notes above.

**Comment by massenz** [comment]
> Are you using Spring Boot 1.5.4.RELEASE?

yes, just realized that may be the issue - upgraded my POM, fixing now some compilation issues that have come up.

I'll update soon as I find out whether that was the problem.

Thanks for super-quick response.

**Comment by massenz** [comment]
Confirmed to work fine with 1.5.4 - however, I had to add the following to my `application.yml` as I was otherwise getting a `401` on the `/prometheus` endpoint (this may be a quirk of my own application configuration; just noting it here in case someone encounters the same issues and google takes him/her here...):
```
# Disabling authentication so that actuator endpoints can be reached.
management:
  security:
    enabled: false
```
Thanks for fixing this, much appreciated!

**Comment by dgradl** [comment]
Oh yeah that security is new in 1.5.4 - you can just disable for the prometheus endpoint as follows:
`
endpoints:
  prometheus:
    sensitive: false
`

**Pull request merged**